Add quickstart guide for Stork Kubernetes Service Discovery (#24821)
quarkusio/quarkus: docs/src/main/asciidoc/stork-kubernetes.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
= Getting Started with SmallRye Stork
:extension-status: preview

> // Temporary until stork in in the BOM

Reviewer: isn't stork in the bom?

Author: I don't know really how it works, I copied it from the `docs/src/main/asciidoc/stork.adoc`

:stork-version: 1.1.0
include::./attributes.adoc[]

The essence of distributed systems resides in the interaction between services.
In modern architecture, you often have multiple instances of your service to share the load or improve the resilience by redundancy.
But how do you select the best instance of your service?
That's where https://smallrye.io/smallrye-stork[SmallRye Stork] helps.
Stork is going to choose the most appropriate instance.
It offers:

* Extensible service discovery mechanisms
* Built-in support for Consul and Kubernetes
* Customizable client load-balancing strategies

include::./status-include.adoc[]

== Prerequisites

:prerequisites-docker:
include::includes/devtools/prerequisites.adoc[]
* Access to a Kubernetes cluster (Minikube is a viable option)

== Architecture

In this guide, we will work with a few components deployed in a Kubernetes cluster:

* A simple blue service.
* A simple red service.
* The `color-service` is the Kubernetes service which is the entry point to the Blue and Red instances.

> * A REST Client calling the blue or red service (the selection is delegated to Stork). The REST Client is also deployed in a namespace within a Kubernetes cluster

Reviewer: These two bullets are about pieces of the same service, right? Shouldn't we merge them? 
Also is the "is also deployed in a namespace..." information important, esp. if we say that they are all in the same namespace a few lines further?

Maybe to something like:
```
* A client service using a REST client to call the blue or the red service. Service discovery and selection are delegated to Stork. The client service is also deployed in the Kubernetes cluster.
```

Author: :+1: Yes good idea. Thanks

* A REST endpoint using the REST client and calling the services

image::stork-kubernetes-architecture.png[Architecture of the application,width=100%, align=center]

For the sake of simplicity, everything will be deployed in the same namespace of the Kubernetes cluster.

== Solution

We recommend that you follow the instructions in the next sections and create the applications step by step.
However, you can go right to the completed example.

Clone the Git repository: `git clone {quickstarts-clone-url}`, or download an {quickstarts-archive-url}[archive].

The solution is located in the `stork-kubernetes-quickstart` {quickstarts-tree-url}/stork-kubernetes-quickstart[directory].

== Discovery and selection

Before going further, we need to discuss discovery vs. selection.

- Service discovery is the process of locating service instances.

> It produces a list of service instances that is potentially empty (if no service matches the request) or contains multiple service instances.

Reviewer: ```suggestion
It produces a list of service instances that is potentially empty (if no service matches the request).
```

if we leave `or contains multiple service instances`, we should probably also add that it can contain a single one too ;)

----
implementation("io.smallrye.stork:stork-service-discovery-kubernetes")
implementation("io.smallrye.stork:stork-load-balancer-random")
implementation("io.quarkus:quarkus-kubernetes")
implementation("io.quarkus:quarkus-kubernetes-client")
implementation("io.quarkus:quarkus-container-image-jib")
----

`stork-service-discovery-kubernetes` provides an implementation of service discovery for Kubernetes. `stork-load-balancer-random` provides an implmentation of random load balancer. `quarkus-kubernetes` enables the generation of Kubernetes manifests each time we perform a build. The `quarkuks-kubernetes-client` extension enables the use of the Fabric8 Kubernetes Client in native mode. And `quarkus-container-image-jib` enables the build of a container image using https://github.com/GoogleContainerTools/jib[Jib].

== The Blue and Red services

Let's start with the very beginning: the service we will discover, select and call.

The Red and Blue are two simple REST services serving an endpoint responding `Hello from Red!` and `Hello from Blue!` respectively. The code of both applications has been developed following the https://quarkus.io/guides/getting-started[Getting Started Guide].

> As the goal of this guide is to show how to use Stork Kubernetes service discovery, we won't provide the code of the Red and Blue services but their container images already built and available in a public registry:

Reviewer: ```suggestion
As the goal of this guide is to show how to use Stork Kubernetes service discovery, we won't provide the specific steps for the Red and Blue services. Their container images are already built and available in a public registry:
```

Author: :+1:

kind: Deployment
metadata:
  annotations:
    app.quarkus.io/commit-id: 27be03414510f776ca70d70d859b33e134570443
    app.quarkus.io/build-timestamp: 2022-03-31 - 10:38:54 +0000
  labels:
    color: red
    type: color-service
    app.kubernetes.io/version: "1.0"
    app.kubernetes.io/name: red-service
  name: red-service
spec:
  replicas: 1
  selector:
    matchLabels:
      app.kubernetes.io/version: "1.0"
      app.kubernetes.io/name: red-service
  template:
    metadata:
      annotations:
        app.quarkus.io/commit-id: 27be03414510f776ca70d70d859b33e134570443
        app.quarkus.io/build-timestamp: 2022-03-31 - 10:38:54 +0000
      labels:
        color: red
        type: color-service
        app.kubernetes.io/version: "1.0"
        app.kubernetes.io/name: red-service
    spec:
      containers:
        - env:
            - name: KUBERNETES_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
          image: quay.io/quarkus/red-service:1.0
          imagePullPolicy: Always
          name: red-service
          ports:
            - containerPort: 8080
              name: http
              protocol: TCP

----

> Note that, as showed by the architecture diagram above, we will have a `color-service` backed by two Pods which are instances of the Red and Blue services. Moreover, the `color-service` is accessible from the outside of the cluster thanks to the Ingress resource.

Reviewer: Just a thought, maybe it would be nicer to use the `<1>`, etc, comments in the code with explanations, as for the rest client reactive guide here: https://quarkus.io/guides/rest-client-reactive#create-the-jax-rs-resource

NOTE: Stork is not limited to Kubernetes and integrates with other service discovery mechanisms.

== The REST Client interface and the front end API

So far, we didn't use Stork; we just deployed the services we will be discovering, selecting, and calling.

We will call the services using the Reactive REST Client.
Create the `src/main/java/org/acme/MyService.java` file with the following content:

[source, java]
----
package org.acme;

import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;

import javax.ws.rs.GET;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

/**
 * The REST Client interface.
 *
 * Notice the `baseUri`. It uses `stork://` as URL scheme indicating that the called service uses Stork to locate and
 * select the service instance. The `my-service` part is the service name. This is used to configure Stork discovery
 * and selection in the `application.properties` file.
 */
@RegisterRestClient(baseUri = "stork://my-service")
public interface MyService {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    String get();
}
----

It's a straightforward REST client interface containing a single method. However, note the `baseUri` attribute.

> It starts with `stork://`.

Reviewer: maybe bullets would be easier to read:
```
It's a straightforward REST client interface containing a single method. However, note the `baseUri` attribute:
* the `stork://` suffix instructs the REST client to delegate the discovery and selection of the service instances to Stork,
* the `my-service` part of the URI is the service name we will be using in the application configuration.
```

package org.acme;

import org.eclipse.microprofile.rest.client.inject.RestClient;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

/**
 * A frontend API using our REST Client (which uses Stork to locate and select the service instance on each call).
 */
@Path("/api")
public class FrontendApi {

    @RestClient MyService service;

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String invoke() {
        return service.get();
    }

}
----

It injects and uses the REST client as usual.

== Stork configuration

Now we need to configure Stork for using Kubernetes to discover the red and blue instances of the service.

In the `src/main/resources/application.properties`, add:

[source, properties]
----
quarkus.stork.my-service.service-discovery.type=kubernetes
quarkus.stork.my-service.service-discovery.k8s-namespace=development
quarkus.stork.my-service.service-discovery.application=color-service
quarkus.stork.my-service.load-balancer.type=random
----

`stork.my-service.service-discovery` indicates which type of service discovery we will be using to locate the `my-service` service.
In our case, it's `kubernetes`.

> If you are connected to a kubernetes cluster, you don't need to configure the access to it. Otherwise, set the proper Kubernetes url using the `quarkus.stork.my-service.service-discovery.k8s-host` property.

Reviewer: ```suggestion
If you are connected to a Kubernetes cluster, you don't need to configure the access to it. Otherwise, set the proper Kubernetes url using the `quarkus.stork.my-service.service-discovery.k8s-host` property.
```

`quarkus.stork.my-service.service-discovery.application` contains the name of the Kubernetes service Stork is going to ask for. In our case, this is the `color-service` corresponding to the kubernetes service backed by the Red and Blue instances.
Finally, `quarkus.stork.my-service.load-balancer.type` configures the service selection. In our case, we use a `random` Load Balancer.

== Deploy the REST Client interface and the front end API in the Kubernetes cluster

> The system is almost complete. We only need to deploy the REST Client interface and the front end API in the cluster.

Reviewer: ```suggestion
The system is almost complete. We only need to deploy the REST Client interface and the front end API to the cluster.
```

In the `src/main/resources/application.properties`, add:

[source, properties]
----
quarkus.container-image.registry=<public registry>
quarkus.kubernetes-client.trust-certs=true
quarkus.kubernetes.ingress.expose=true
quarkus.kubernetes.ingress.host=my-service.127.0.0.1.nip.io
----

The `quarkus.container-image.registry` contains the container registry to use.
The `quarkus.kubernetes.ingress.expose` indicates that the service will be accessible from the outside of the cluster.
The `quarkus.kubernetes.ingress.host` contains the url to access the service. We are using https://nip.io/[nip.io] wildcard for IP address mappings.

For a more customized configuration you can check the https://quarkus.io/guides/deploying-to-kubernetes[Deploying to Kubernetes guide]

== Build and push the container image

> Thanks to the extensions we are using, we enable the build of a container image using Jib while also enabling the generation of Kubernetes manifests each time we perform a build. For example, the following command will generate a Kubernetes manifest in the `target/kubernetes/` directory and also build and push a container image for the project:

Reviewer: ```suggestion
Thanks to the extensions we are using, we can perform the build of a container image using Jib and the generation of Kubernetes manifests while building the application. For example, the following command will generate a Kubernetes manifest in the `target/kubernetes/` directory and also build and push a container image for the project:
```

[source,shell script]
----
./mvnw package -Dquarkus.container-image.build=true -Dquarkus.container-image.push=true
----

> == Deploy REST Client interface and the front end API in the Kubernetes cluster

Reviewer: ```suggestion
== Deploy the client service to the Kubernetes cluster
```
